metrics_test.go

```diff
@@ -0,0 +1,204 @@
+		num := 10
+		val := make([]byte, 1<<20) // Large Value
+		key := make([]byte, 40)
+		for i := 0; i < num; i++ {
+			_, err := rand.Read(key)
+			require.NoError(t, err)
+			_, err = rand.Read(val)
+			require.NoError(t, err)
+
+			writer := db.NewManagedWriteBatch()
+			require.NoError(t, writer.SetEntryAt(NewEntry(key, val), 1))
+			writer.Flush()
+		}
+
+		expectedSize := int64(len(val)) + 200 // vlog expected size
+
+		totalWrites := expvar.Get("badger_v4_vlog_writes_num")
+		require.Equal(t, int64(num), totalWrites.(*expvar.Int).Value())
+
+		bytesWritten := expvar.Get("badger_v4_vlog_written_bytes")
+		require.GreaterOrEqual(t, expectedSize*int64(num), bytesWritten.(*expvar.Int).Value())
+
+		txn := db.NewTransactionAt(2, false)
+		item, err := txn.Get(key)
+		require.NoError(t, err)
+		require.Equal(t, uint64(1), item.Version())
+
+		err = item.Value(func(val []byte) error {
+			totalReads := expvar.Get("badger_v4_vlog_reads_num")
+			bytesRead := expvar.Get("badger_v4_vlog_reads_bytes")
+			require.Equal(t, int64(1), totalReads.(*expvar.Int).Value())
+			require.GreaterOrEqual(t, expectedSize, bytesRead.(*expvar.Int).Value())
+			return nil
+		})
+
+		require.Nil(t, err)
+	})
+}
+
+func ClearAllMetrics() {
+	expvar.Do(func(kv expvar.KeyValue) {
+		// Reset the value of each expvar variable based on its type
+		switch v := kv.Value.(type) {
+		case *expvar.Int:
+			v.Set(0)
+		case *expvar.Float:
+			v.Set(0)
+		case *expvar.Map:
+			v.Init()
+		case *expvar.String:
+			v.Set("")
+		}
+	})
+}
+
+func TestReadMetrics(t *testing.T) {
+	opt := getTestOptions("")
+	opt.managedTxns = true
+	opt.CompactL0OnClose = true
+	runBadgerTest(t, &opt, func(t *testing.T, db *DB) {
+		ClearAllMetrics()
+		num := 10
+		val := make([]byte, 1<<15)
+		keys := [][]byte{}
+		writer := db.NewManagedWriteBatch()
+		for i := 0; i < num; i++ {
+			keyB := key("byte", 1)
+			keys = append(keys, []byte(keyB))
+
+			_, err := rand.Read(val)
+			require.NoError(t, err)
+
+			require.NoError(t, writer.SetEntryAt(NewEntry([]byte(keyB), val), 1))
+
+		}
+		writer.Flush()
+
+		txn := db.NewTransactionAt(2, false)
+		item, err := txn.Get(keys[0])
+		require.NoError(t, err)
+		require.Equal(t, uint64(1), item.Version())
+
+		totalGets := expvar.Get("badger_v4_gets_total")
+		require.Equal(t, int64(1), totalGets.(*expvar.Int).Value())
+
+		totalMemtableReads := expvar.Get("badger_v4_memtable_gets_total")
+		require.Equal(t, int64(1), totalMemtableReads.(*expvar.Int).Value())
+
+		totalLSMGets := expvar.Get("badger_v4_lsm_level_gets_total")
+		require.Nil(t, totalLSMGets.(*expvar.Map).Get("l6"))
+
+		// Force compaction
+		db.Close()
+
+		db, err = OpenManaged(opt)
+		require.Nil(t, err)
+
+		txn = db.NewTransactionAt(2, false)
+		item, err = txn.Get(keys[0])
+		require.NoError(t, err)
+		require.Equal(t, uint64(1), item.Version())
+
+		_, err = txn.Get([]byte(key("abdbyte", 1000))) // val should be far enough that bloom filter doesn't hit
+		require.NotNil(t, err)
```

Comment:
for errors, we should avoid using NotNil/Nil. It doesn't print well when it fails.